src/util/heap.ts

```diff
@@ -51,6 +51,14 @@ export class Heap<K, V> {
     return this.nodes[0];
   }
 
+  public release(node: HeapNode<K, V>): void {
+    for (const _node of this.nodes) {
+      if (_node.getValue() !== node.getValue()) {
+        this.nodes.push(node);
+      }
+    }
+  }
```

Comment:
In Go, [PriorityQueue.Release](https://github.com/yorkie-team/yorkie/blob/master/pkg/pq/priority_queue.go#L53) was intended to delete the given node from the heap. It was difficult to find and delete the target node from the heap, so we added all nodes to the new heap excluding that node.

I think we need to do similar processing in JavaScript.

Reply:
I would like to suggest a more efficient way if it is simply changing the nodes array.
It seems that you can delete only certain nodes without creating a new list and going through all the loops.
example ::
```js
// heap.ts
export class HeapNode<K, V> {
...
public releaseByFindIdx(node: HeapNode<K, V>): void {
    const idx = this.nodes.findIndex((_node) => {
      return _node.getValue() === node.getValue();
    });

    if (idx > -1) {
      this.nodes.splice(idx, 1);
    }
  }
...
}
```

```js
// heap_test.ts
...
it('Release performance comparison test', function () {
    let removeNode = null;

    function createHeap(): Heap<number, number> {
      const size = 10000000;
      const heap = new Heap<number, number>();

      for (const idx of shuffle(range(0, size))) {
        const node = new HeapNode(idx, idx);
        heap.push(node);
        if (idx === size - 1) {
          removeNode = node;
        }
      }
      return heap;
    }

    const heap1 = createHeap();
    let start = +new Date();
    heap1.release(removeNode);
    let end = +new Date();
    logger.debug(`used release : ${end - start}`);

    const heap2 = createHeap();
    start = +new Date();
    heap2.releaseByFindIdx(removeNode);
    end = +new Date();
    logger.debug(`used releaseByFindIdx : ${end - start}`);

    // for test
    assert.equal(0, 1);
  });
...
```

```bash
# result
LOG: 'YORKIE D:', 'used release : 467'
Chrome Headless 86.0.4240.193 (Mac OS 10.15.7): Executed 22 of 45 SUCCESS (0 secs / 0.609 secs)
LOG: 'YORKIE D:', 'used releaseByFindIdx : 16'
...
```
Looking at the results, I think it is better to use `findIndex` and `splice` in terms of performance and memory usage.
I'm curious about your opinion. @hackerwins @mojosoeun 

[`findIndex` vs `for loop`] https://www.measurethat.net/Benchmarks/Show/9343/0/object-arrays-findindex-vs-for-loop-with-logging

Reply:
It seems that my previous comments were not sufficiently explained.

`heap.ts` is a [binary heap](https://en.wikipedia.org/wiki/Binary_heap#:~:text=A%20binary%20heap%20is%20a,a%20data%20structure%20for%20heapsort.) implemented with an array. The index of the element in the array represents the parent-child relationship, so it is difficult to delete it like a normal list when deleting.

I haven't looked at it in detail, but it seems that how to delete a specific node from the heap is described below.
http://www.mathcs.emory.edu/~cheung/Courses/171/Syllabus/9-BinTree/heap-delete.html

Reply:
Thank you for correcting my mistakes. :)
It seems to me that it can be solved using the methods of the class. Is that right?

Reply:
That's right.

When I tried `deleting a node in a heap` in Go, it was difficult to delete a node [in-place way](https://en.wikipedia.org/wiki/In-place_algorithm), so I implemented it by creating a new heap and adding nodes excluding the target node.

Reply:
I'm thinking about implementing like this and add a test for release function in the heap_test.ts 
```javascript
public release(node: HeapNode<K, V>): void {
    for (const [i, _node] of this.nodes.entries()) {
      if (_node.getValue() === node.getValue()) {
        this.nodes[i] = this.nodes.pop();
        const parentIndex = this.getParentIndex(i);

        if (parentIndex > -1) {
          this.heapify(this.nodes[this.getParentIndex(i)], this.nodes[i], i);
        }
        break;
      }
    }
 }

public heapify(
  parentNode: HeapNode<K, V>,
  currentNode: HeapNode<K, V>,
  targetIndex: number,
 ): void {
  if (this.comparator(parentNode.getKey(), currentNode.getKey()) < 0) {
    this.moveUp(targetIndex);
  } else {
    this.moveDown(targetIndex);
  }
}
```
If we assume heap nodes is [9, 8, 5, 6, 7, 1, 4, 0, 3, 2], and if we want to delete `5`, we will get like this [9, 8, 4, 6, 7, 1, 2, 0, 3]

Reply:
I'm thinking about implementing like this and add a test for release function in the heap_test.ts 
```javascript
public release(node: HeapNode<K, V>): void {
    for (const [i, _node] of this.nodes.entries()) {
      if (_node.getValue() === node.getValue()) {
        this.nodes[i] = this.nodes.pop();
        this.heapify(this.nodes[this.getParentIndex(i)], this.nodes[i], i);
        break;
      }
    }
 }

public heapify(
  parentNode: HeapNode<K, V>,
  currentNode: HeapNode<K, V>,
  targetIndex: number,
 ): void {
  if (parentNode &&this.comparator(parentNode.getKey(), currentNode.getKey()) < 0) {
    this.moveUp(targetIndex);
  } else {
    this.moveDown(targetIndex);
  }
}
```
If we assume heap nodes is [9, 8, 5, 6, 7, 1, 4, 0, 3, 2], and if we delete `5`, we will get like this [9, 8, 4, 6, 7, 1, 2, 0, 3]

Reply:
Cool. It was easier to solve than I thought. I think it would be nice to put `heapify` in heap.ts. After this PR merge, it would be nice to add `heapify` to Go. :)